hphp/runtime/vm/jit/code-cache.cpp

```diff
@@ -97,7 +98,14 @@ CodeCache::CodeCache()
     exit(1);
```

Comment:
Whoa, sorry @jim-saxman; I must have missed the notification (or maybe it got dropped) for this comment.

That seems fine to me, although if the user configures HHVM to have larger code sizes than the defaults (which I suspect we may do internally), then you'll still run into the performance degradation with far jumps.  There might be reasons not do to it that I'm not aware of, but if there are, someone will raise them once we see the PR.

+ @dave-estes, @cmuellner, @swalk-cavium—On the subject of avoiding long jumps: I think the most important thing is for hot and main to be within near jump distance of one another.  Jumps to cold and frozen are supposed to be rare, so the cost penalty should be relatively insignificant in practice—though, of course, the far jump is a much bigger instruction sequence, which does still have a passive impact on icache behavior.

One idea we bounced around internally when discussing this issue on ARM was to dynamically allocate blocks of far jump trampolines which were always within a near jump of hot and main.  Then, in the code we emit, we can just _always_ emit a near jump, and then proxy through the trampoline if necessary.  The static relocator could probably handle this, and it would obviate the dependency on dynamic relocation (though smashable jumps would still require dynamic relocation to shorten them, I'd expect).  (Are calls PC-relative and limited in immediate size, also?  Even if main itself grew too big, we could partition main into chunks with trampoline areas in between, and the same solution would work if we called across chunks in main.)

This is an area with substantial potential for perf improvement, so I'm interested in hearing other thoughts or ideas on the subject you all might have.